Xamarin.PropertyEditing.Mac/Controls/Custom/AutocompleteTextField.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using AppKit;
+using Foundation;
+using Xamarin.PropertyEditing.ViewModels;
+
+namespace Xamarin.PropertyEditing.Mac
+{
+	internal class AutocompleteTextField : NSTextField
+	{
+		private const string AutocompleteItemsString = "AutocompleteItems";
+		private const string PreviewCustomExpressionString = "PreviewCustomExpression";
+
+		private IReadOnlyList<string> autocompleteValues;
+		private PropertyInfo previewCustomExpressionInfo;
+		private PropertyInfo customcompleteItemsPropertyInfo;
+
+		private PropertyViewModel viewModel;
+		public PropertyViewModel ViewModel {
+			get { 
+				return this.viewModel; 
+			}
+			internal set {
+				this.viewModel = value;
+
+				if (this.viewModel != null) {
+					var vmType = ViewModel.GetType ();
+					this.customcompleteItemsPropertyInfo = vmType.GetProperty (AutocompleteItemsString);
+
+					if (this.previewCustomExpressionInfo == null)
+						this.previewCustomExpressionInfo = vmType.GetProperty (PreviewCustomExpressionString);
+
+					(Delegate as AutocompleteTextFieldDelegate).TextChanged += (object sender, string e) => {
+						this.previewCustomExpressionInfo.SetValue (this.viewModel, e);
+					};
+				}
+			}
+		}
+
+		public AutocompleteTextField ()
+		{
+			Delegate = new AutocompleteTextFieldDelegate ();
+		}
+
+		internal string[] Suggestions ()
+		{
+			var values = customcompleteItemsPropertyInfo.GetValue (this.viewModel);
+			if (values != null && this.autocompleteValues == null)
+				this.autocompleteValues = ((IReadOnlyList<string>)values);
+
+			return autocompleteValues.ToArray ();
+		}
+	}
+
+	internal class AutocompleteTextFieldDelegate : NSTextFieldDelegate
```

Comment:
This could be a child class to `AutocompleteTextField`. Nothing else will need to interact with it, and then `Suggestions()` (which should be `GetSuggestions()`) can be private.